### DISCUSSION TITLE
Communication protocol simplification possible?

### DISCUSSION THREAD

**Discussion by ralphwetzel** [phoddie/node-red-mcu#30] [Ideas]
@phoddie --
Currently the status message consists of the status data (necessary for `node-red`) and an additional object that provides `source` information. The (only?) relevant property of the `source` object is the `id` of the emitting node, that is used by the plugin to build a status notification that is understood by the editor.

https://github.com/phoddie/node-red-mcu/blob/7ddfd04e5052f8a8859e0c7db7597fd484b327f1/nodered.js#L296-L306

Thus `type` and `name` are not used & can (as well) be derived from the provided `id` by the node-red runtime. They yet need to be processed & transmitted with each status message - adding a significant load without value.

I propose to change the protocol to just provide the node's `id` & omit either `type` and `name`. I'm fine with keeping the property named as `source`, whereas `id` would fit as well:

```javascript
 status(status) { 
 	const msg = { 
 		status: { 
 			...status, 
 			source: this.id
 		} 
 	}; 
```

This proposal is of course valid not only for `status`, but for all messages `trace`d back to the `node-red` runtime.

**Comment by phoddie** [comment]
To start, I think we should review the message and how it is used. The message is a match for the message that Node-RED sends to Status Nodes when the status is changed. Here's a flow to explore that.

<img width="748" alt="image" src="https://user-images.githubusercontent.com/6856458/198857748-ba745ed5-51ed-4d66-866a-1764fea75b62.png">

<details>

```json
[
    {
        "id": "2fec9e130ad0d45f",
        "type": "tab",
        "label": "delete me",
        "disabled": false,
        "info": "",
        "env": []
    },
    {
        "id": "5e3247acce2fd674",
        "type": "status",
        "z": "2fec9e130ad0d45f",
        "name": "",
        "scope": null,
        "_mcu": false,
        "x": 480,
        "y": 220,
        "wires": [
            [
                "16b7efbf68b385c1"
            ]
        ]
    },
    {
        "id": "16b7efbf68b385c1",
        "type": "debug",
        "z": "2fec9e130ad0d45f",
        "name": "debug 8",
        "active": true,
        "tosidebar": false,
        "console": true,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": false,
        "x": 670,
        "y": 220,
        "wires": []
    },
    {
        "id": "1790bba3c0262e5c",
        "type": "inject",
        "z": "2fec9e130ad0d45f",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": true,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "_mcu": false,
        "x": 170,
        "y": 360,
        "wires": [
            [
                "9afeee4e3f8febbb"
            ]
        ]
    },
    {
        "id": "9afeee4e3f8febbb",
        "type": "function",
        "z": "2fec9e130ad0d45f",
        "name": "function 3",
        "func": "node.status({fill:\"red\",shape:\"ring\",text:\"disconnected\"});",
        "outputs": 0,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "_mcu": false,
        "x": 340,
        "y": 360,
        "wires": []
    }
]
```

</details>

In Node-RED, it outputs the following to the console:

```text
29 Oct 17:45:37 - [info] [debug:debug 8] 
{
  status: {
    fill: 'red',
    shape: 'ring',
    text: 'disconnected',
    source: { id: '9afeee4e3f8febbb', type: 'function', name: 'function 3' }
  },
  _msgid: '6c17a86e265510f2'
}
```

Using Node-RED MCU it outputs the following to xsbug (I've reformatted it for readability). 
```json
{
	"status": {
		"fill": "red",
		"shape": "ring",
		"text": "disconnected",
		"source": {
			"id": "9afeee4e3f8febbb",
			"type": "function",
			"name": "function 3"
		}
	},
	"_msgid": "287b8cad8de8ef90"
}
```

This same message object is also sent to xsbug to relay to the Node-RED Editor. Changing the message propagated to Status Nodes will break compatibility with Node-RED.

The message sent to xsbug could be simplified by removing unnecessary fields - in addition to the source `type` and `name`, it seems possible that the `_msg` ID is also unnecessary. But that would need to be done in such a way that the messages sent to Status Nodes maintain compatiblity.

As an aside, you mention "adding a significant load without value". I'm curious if you are able to quantify the load impact?

**Reply by ralphwetzel** [reply]
> As an aside, you mention "adding a significant load without value". I'm curious if you are able to quantify the load impact?

``` xml
<xsbug>
<bubble name="" value="1" path="/Users/ralph/Projekte/node-red-mcu-plugin/node-red-mcu/nodered.js" line="289">{&#34;status&#34;:{&#34;fill&#34;:&#34;blue&#34;,&#34;shape&#34;:&#34;dot&#34;,&#34;text&#34;:1,&#34;source&#34;:{&#34;id&#34;:&#34;43232f662ac1fdac&#34;,&#34;type&#34;:&#34;Node-RED Compatibility&#34;,&#34;name&#34;:&#34;upto 5s&#34;}}}
</bubble>
</xsbug>
```
This is an example of a `status` trace as it is currently sent from the MCU to xsbug (and caught in-between by the node-red runtime). It's length is 417 chars. Without the properties `source.name` and `source.type`, it's 81 chars less. That's a bit more than 19% of bandwidth invested to transmit data that is not used.

Of course this is just one message & a bigger sample would be necessary to get fully representative averages - and yet it already is significant.

**Reply by ralphwetzel** [reply]
> The message is a match for the message that Node-RED sends to Status Nodes when the status is changed.

It [looks like Node-RED adds necessary data](https://github.com/node-red/node-red/blob/946def022fa94e9998d5c6095838841a1c94e2da/packages/node_modules/%40node-red/runtime/lib/flows/Flow.js#L570-L630) while processing the status message:
* Initially, it emits a `node-status` event (591), that carries just the node's `id` and the status information as data.
* If there are (!!) any `StatusNode`s, it then amends the `source` object (619) and forwards this pack as `msg`.

This is exactly what I had in mind when opening this discussion.

There's another aspect & perhaps a potential for further improvement: The `source` properties never change (for a dedicated node). Yet currently this object it is newly built & processed by `JSON.stringify` for each status message - consuming resources. Couldn't `nodered2mcu` preprocess this & _inject_ it adequately?

**Reply by phoddie** [reply]
A change is committed to the implementation of `Node.prototype.status` to omit the `type` and `name` fields. 

> There's another aspect & perhaps a potential for further improvement: The source properties never change (for a dedicated node). Yet currently this object it is newly built & processed by JSON.stringify for each status message - consuming resources. Couldn't nodered2mcu preprocess this & inject it adequately?

The message passed to `JSON.stringify` includes the status, which is not guaranteed to be a constant. Consequently, I don't see how this could be optimized at build-time.

**Reply by ralphwetzel** [reply]
> A change is committed to the implementation of `Node.prototype.status` to omit the `type` and `name` fields.

Thank you! 👍 

I would like to propose to even simplify one step further, using a feature of `trace.left`:

``` javascript
	status(status) {
		let msg = {
			status: {
				...status
			}
		};

		trace.left(JSON.stringify(msg), this.id);

		const statuses = this.#outputs.statuses;
		if (!statuses)
			return;
		
		msg.source = {
			id: this.id,
			type: this.constructor.type,
			name: this.name
		}

		for (let i = 0, length = statuses.length; i < length; i++)
			RED.mcu.enqueue(msg, statuses[i]);
	}
```

**Reply by ralphwetzel** [reply]
Speaking w/ data: The total of all proposed simplification shortens the message shown above from 417 chars to 253 chars; 40% less data to process & transmit.

```xml
<xsbug>
<bubble name="43232f662ac1fdac" value="1" path="/Users/ralph/Projekte/node-red-mcu-plugin/node-red-mcu/nodered.js" line="289">{&#34;status&#34;:{&#34;fill&#34;:&#34;blue&#34;,&#34;shape&#34;:&#34;dot&#34;,&#34;text&#34;:1}}
</bubble>
</xsbug>
```

**Reply by phoddie** [reply]
It is an interesting idea to use the optional `conversation` argument of `trace.left` to carry the node ID. That allows filtering in the xsbug message pane. It could be OK, but first I have a few questions / observations.

1. Is it meaningfully smaller? I think it does save a few bytes.
2. It makes the message sent by Node-RED MCU Edition more unlike the message in Node-RED. Generally I've tried to keep them the same to minimize the chance to translation errors. That's for you to manage. ;)
3. If this change is made, corresponding change to the Debug Node's `onMessage` method should also be made so the protocol is consistent. Correct?
4. The biggest win remaining to reduce the payload size is be to suppress the `path` and `line` attributes, but that would require deeper changes and potentially modifying xsbug too.

**Reply by ralphwetzel** [reply]
> 2\. That's for you to manage. ;)

👍 A bit more effort for the plugin - but no big deal, and already done.

> 3\. If this change is made, corresponding change to the Debug Node's `onMessage` method should also be made so the protocol is consistent. Correct?

Yes. That makes sense.

> 4\. The biggest win remaining to reduce the payload size is be to suppress the `path` and `line` attributes, but that would require deeper changes and potentially modifying xsbug too.

This is going to lead to a dedicated protocol for the Node-RED communication. Whereas I confirm the additional potential, I have the impression, it's - currently - not demanded. I'll do some further stress-testing the next days; perhaps this gives us any supportive data...

**Reply by phoddie** [reply]
I've  committed a small change to set the conversation to the node ID on the messages for the editor (debug and status). Because this is a breaking change to the protocol, I did not remove the `id` property from the `source` yet. Once you have updated the plug-in to depend only on the conversation value as the ID, let me know and I'll remove those. I also have some work planned to improve the debug node output that I'll do at that point.

> This is going to lead to a dedicated protocol for the Node-RED communication

FWIW – I don't think it needs to go that far. The xsbug protocol has proven to be remarkably robust and flexible. That's good, because it is also deeply intertwined with XS itself, so it isn't practical to make significant changes.

I would like to propose one additional change that will require support from the plug-in. When flows are running and the editor is **not** attached, the MCU is still preparing and sending messages to the editor. That's unnecessary. A lightbulb running Node-RED won't always be connected to the editor. ;) So... I'd like to propose that the plug-in sends a message to Node-RED on the MCU when it connects. That will be a signal to the Node-RED MCU runtime that it should send messages for the editor. That message can be delivered using the same mechanism as "inject". The payload can simply be an empty object (which will be ignored now because it has no `flow` or `Id`). What do you think about adding that to the plugin?

**Reply by ralphwetzel** [reply]
> Once you have updated the plug-in to depend only on the conversation value as the ID, let me know and I'll remove those.

Done now. The plugin understands both protocols.

> the plug-in sends a message to Node-RED on the MCU when it connects

Of course. I would even go a bit further: Let's change the default status from "MCU sends data." (as it does today) to "MCU doesn't send data." Then it's up to an external trigger to motivate the MCU to change into sending state. One message of the plugin could e.g. be worth 60 seconds of data sent by the MCU. Once this time is over, the MCU has either received another external trigger (restarting the 60s period) - or falls back into non-sending mode.

**Comment by phoddie** [comment]
> Done now. The plugin understands both protocols.

Cool. Thanks. I will try to updated the MCU runtime for that this evening.

> Let's change the default status from "MCU sends data." (as it does today) to "MCU doesn't send data." Then it's up to an external trigger to motivate the MCU to change into sending state.

Exactly. The empty message from your plug-in is a signal to send the state to the editor.

> One message of the plugin could e.g. be worth 60 seconds of data sent by the MCU. Once this time is over, the MCU has either received another external trigger (restarting the 60s period) - or falls back into non-sending mode

This is additional complexity and a bit of memory for a relatively obscure situation. Let's hold onto this idea. If there's a clear need, it can be implemented then.

**Reply by phoddie** [reply]
A bunch of changes committed related to the protocol and the Debug node:
- `id` only passed as conversation ID. `source` completely eliminated
- When debug node sends message to status, truncates to 32-characters (like Node-RED)
- xsbug conversation only used for messages to Node-RED editor
- Debug console and side-bar message both sent to xsbug log. Uses new colorize feature to distinguish console.log, debug console, and debug log messages. More usable that before.
- Debug node implements active (suppress output in Node-RED Editor)
- Debug node implements `onCommand` to allow MCU plug-in to turn `active` on/off (saves bandwidth & time when off!)

It is a little complex to test all of this carefully, so there may be errors and oversights. At a minimum, this should be a good step to build on. Please let me know how it goes.

**Reply by ralphwetzel** [reply]
> * `id` only passed as conversation ID. `source` completely eliminated

Confirmed working! 👍 

> * Debug node implements active (suppress output in Node-RED Editor)
> * Debug node implements `onCommand` to allow MCU plug-in to turn `active` on/off (saves bandwidth & time when off!)

The way this is currently implemented demands the plugin to trigger each DebugNode individually. I'd rather setup one global flag to enable / disable the communication with the Node-RED runtime. Additionally, it might be that the plugin thinks it's enabling the comm - but it is switching it off (as sending a second trigger disables #active).

Let me come back to my earlier proposal 😉 :

```javascript
let debug_active;

globalThis["<xsbug:script>"] = function(mystery, path, line, script) {
    const options = JSON.parse(script);

    if (options.ping) {
        if (!debug_active) {
            debug_active = Timer.set(() => {
                debug_active = undefined;
            }, 60000);
        } else {
            Timer.schedule(debug_active, 60000);
        }
        return;
    }

    const node = flows.get(options.flow)?.getNode(options.id);
    if (!node)
        return;
    node.onCommand(options);
}

```

The 'cost' is one timer, one global variable and one `JSON.parse` operation once per minute. A `ping` always enables the communication; no ambiguity. The flow falls back (quite soon) into non-sending mode in the absence of a timely `ping`.

**Reply by phoddie** [reply]
Thanks for confirming that the ID related changes are working. That's a nice simplification.

I think I may not have explained myself clearly regarding the `active` capability of the Debug node. Apologies. The capability is there to support the ability to dynamically turn Debug output on and off in the Node-RED Editor. Here's a screen recording that shows the Editor behavior that it supports.

> https://user-images.githubusercontent.com/6856458/199794663-61a74c45-cb6e-40f8-9bdc-a05062f76ff9.mov

**Reply by ralphwetzel** [reply]
Ok & sorry! The misunderstanding is on my side. Now I got it; my intension was a completely different one. 
Let's see, what's necessary to implement this feature...

**Reply by ralphwetzel** [reply]
> Let's see, what's necessary to implement this feature...

Two topics I'd like to mention already:
* There's a `data` property in `options` that was foreseen to transmit command parameters.

https://github.com/phoddie/node-red-mcu/blob/93c9f6758c0ff8da43c50ed953882c80f94804f0/nodered.js#L489-L492

Line 491: `this.#active = !!options.data;`

* I became a bit confused as the debug messages continued to appear in the `xsbug Console` - despite being switched off in the editor. Wouldn't it make sense to consistently disable these messages - neither display them in xsbug, nor send them back to NR?

**Reply by ralphwetzel** [reply]
The functionality should be available with [the latest commit](https://github.com/ralphwetzel/node-red-mcu-plugin/commit/1f6dd20e1ecc85a7d659e5de7764b73aff6011cf). This yet requires the small change in `line 491`, as shown.

**Reply by phoddie** [reply]
Nice. I've pushed the change from `active` to `data`.

> I became a bit confused as the debug messages continued to appear in the xsbug Console - despite being switched off in the editor. Wouldn't it make sense to consistently disable these messages - neither display them in xsbug, nor send them back to NR?

I could go either way on the behavior. Since continuing to log in xsbug confused you, let's try suppressing those for now to match the Editor.

**Reply by phoddie** [reply]
BTW – it is pretty cool that the buttons of both the Inject and Debug nodes in the Editor are handled by the MCU. Nice work.
